### ISSUE TITLE
Const patterns can have different behavior from PartialEq, due to incorrect bounds in `derive(PartialEq)`

### ISSUE THREAD

**Issue by theemathas** [rust-lang/rust#147714]
I tried this code:

```rust
#[allow(dead_code)]
#[derive(PartialEq)]
enum Thing<T> {
    A(T),
    B,
}

struct Incomparable;

impl PartialEq for Thing<Incomparable> {
    fn eq(&self, _: &Self) -> bool {
        panic!()
    }
}

const X: Thing<Incomparable> = Thing::B;

fn main() {
    if let X = X {
        println!("equal");
    }
}
```

I expected the code to either produce a compile error, or panic at run time. Instead, it compiled and printed "equal". That is, `if let X = X` has different behavior from `if X == X`.

This seems to be caused by the `derive(PartialEq)` expanding to the following code, which seems incorrect, due to the lack of bounds on `T`:
```rust
impl<T> ::core::marker::StructuralPartialEq for Thing<T> { }
```

[The reference](https://doc.rust-lang.org/1.90.0/reference/patterns.html#r-patterns.const.aggregate) seems to be ambiguous on whether this behavior is correct or not, but I think that this behavior is incorrect, since I interpret "its `PartialEq` instance" to mean "the impl that caused the `const`'s type to implement `PartialEq`".

### Meta

Reproducible on the playground with version `1.92.0-nightly (2025-10-13 4b94758d2ba7d0ef71cc)`

**Comment by BoxyUwU** [comment]
100% a bug. we don't intend for const patterns to be observably different from a `PartialEq` call

**Comment by BoxyUwU** [comment]
I guess this is jsut caused by `StructuralPartialEq` not inheriting the `<T>` parameter on `PartialEq`. In theory deriving `PartialEq<T>` should implement `StructuralPartialEq<T>`. not a blanket "all partial eq impls are `StructuralPartialEq`
